### PR TITLE
#2107 - E2E test for offering bulk upload

### PR DESCRIPTION
**Pull request by ann-aot**
- Created 2 TCs for the bulk insert.
image
-  I have created a function to clear test db data which is created as part of the TCs because, when a new location or program is created that matches with the CSV for each TC, Due to the unique constraints in DB, we need to clear it.
-  Updated `createFakeEducationProgram` and `createFakeInstitutionLocation`

### CODE REVIEW

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+
+  it(
+    "Should create an approved and a creation pending offerings when a multi line bulk" +
+      " offering CSV file with existing location code and SABC code, and one with" +
+      " same delivery method and another with different delivery method is uploaded.",
+    async () => {
+      // Arrange
+      // Creating an institution location with same location code as that of the
+      // first row of the multiple CSV file.
+      const collegeFLocationYESK = createFakeInstitutionLocation(
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:91]: The API test suits run in parallel, hence there is no guarantee that the location with institution code YESK (in this case) could not be created before. 

I had the same situation while creating e2e tests for ece response integration. I followed this approach and it worked.
image

We could possibility centralize and re-use the same method. Let me know if I am missing something.

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+
+      await authorizeUserTokenForLocation(
+        db.dataSource,
+        InstitutionTokenTypes.CollegeFUser,
+        collegeFLocationYESK,
+      );
+
+      // Act/Assert
+      await request(app.getHttpServer())
+        .post(`${endpoint}?validation-only=true`)
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:241]: 👍

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+        csvProgramSABCCodeSBC2,
+        collegeFLocationYESK,
+        csvLocationCodeYESK,
+      );
+    },
+  );
+
+  it("Should return program related validation error when bulk offering CSV file with a non existing program SABC code is uploaded. ", async () => {
+    // Arrange
+    const randomSABCCode = `XXXX1`;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:287]: Template literals not required.

File: sources/packages/backend/apps/api/src/testHelpers/fake-entities/education-program-fake.ts [hunk 4]
```diff
@@ -8,20 +8,34 @@ import {
...
 import { createFakeInstitution } from "@sims/test-utils";
 
+/**
+ * Create fake education program.
+ * @param relations dependencies.
+ * - `institution` related institution.
+ * - `user` related user.
+ * @param options dependencies.
+ * - `initialValues` initial values.
+ * @returns
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/testHelpers/fake-entities/education-program-fake.ts:18]: comment for returns.

**Comment by dheepak-aot** [comment]
Great Job. Thanks for doing the refactor of `createFakeInstitutionLocation`. Other than the discussion we had on cleanTestData, just 2 minor comments.

File: sources/packages/backend/apps/api/src/testHelpers/fake-entities/education-program-fake.ts [hunk 5]
```diff
@@ -32,9 +46,9 @@ export function createFakeEducationProgram(
   program.hasTravel = "yes";
   program.hasIntlExchange = "yes";
   program.programDeclaration = true;
-  program.institution = institution ?? createFakeInstitution();
+  program.institution = relations?.institution ?? createFakeInstitution();
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/testHelpers/fake-entities/education-program-fake.ts:49]: 👍

**Review by guru-aot — APPROVED** [review]
Nice work @ann-aot

File: sources/packages/backend/libs/test-utils/src/factories/institution-location.ts [hunk 6]
```diff
@@ -4,16 +4,25 @@ import { createFakeInstitution } from "./institution";
...
- * @param institution institution.
+ * @param relations dependencies.
+ * - `institution` related institution.
+ * @param options dependencies.
+ * - `initialValues` initial values.
  * @returns institution location
  */
 export function createFakeInstitutionLocation(
-  institution?: Institution,
+  relations?: {
```
Comment (sh16011993) [inline sources/packages/backend/libs/test-utils/src/factories/institution-location.ts:14]: 👍

File: sources/packages/backend/apps/api/src/testHelpers/fake-entities/education-program-fake.ts [hunk 7]
```diff
@@ -8,20 +8,34 @@ import {
...
+ * - `institution` related institution.
+ * - `user` related user.
+ * @param options dependencies.
+ * - `initialValues` initial values.
+ * @returns
+ */
 export function createFakeEducationProgram(
-  institution?: Institution,
-  user?: User,
+  relations?: {
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/testHelpers/fake-entities/education-program-fake.ts:21]: 👍

**Review by sh16011993 — APPROVED** [review]
LGTM 👍

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+        InstitutionTokenTypes.CollegeFUser,
+        collegeFLocationKSEY,
+      );
+
+      const multipleOfferingFilePath = path.join(
+        __dirname,
+        "bulk-insert/multiple-upload.csv",
+      );
+
+      let [responseOfferingSBC1, responseOfferingSBC2] = [undefined, undefined];
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:144]: Suggesttion: variables are implicitly undefined when not initialized. 
We can use like this.
```ts
let responseOfferingSBC1: EducationProgramOffering,
        responseOfferingSBC2: EducationProgramOffering;
```

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+        });
+
+      // Checking the created offering statuses.
+      const [offeringSBC1, offeringSBC2] =
+        await db.educationProgramOffering.find({
+          select: {
+            offeringStatus: true,
+          },
+          where: {
+            id: In([responseOfferingSBC1.id, responseOfferingSBC2.id]),
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:164]: Shouldn't we use the order by id to guarantee the result order?

**Review by dheepak-aot — APPROVED** [review]
Thanks for doing the changes 👍

**Pull request merged**